### DISCUSSION TITLE
How to read in an UnstructuredGrid from FEA

### DISCUSSION THREAD

**Discussion by AstaBlak** [pyvista/pyvista#4909] [Q&A]
Good Day Team, 

I have recently been learning python coding and learning to use the pyvista library. Since I am new to pyvista, I have been trying to follow through with the examples to understand using the library. Still, the learning process is somewhat confusing for someone who isn't familiar with VTK or pyvista. 

I need some help or guidance with trying to achieve my project.

Simple Scope:
I am simply trying to create a final model (preferably an stl or obj file format output) based on X, Y, Z coordinate points exported from an FEA model.  

Reasoning: 
My idea is to build a new model from specific x,y,z, coordinate points, which are extracted from the FEA software based on a stress range. See the example image below: 
![Exporting_stress_range](https://user-images.githubusercontent.com/85831923/121809566-98a66a80-cc5d-11eb-916f-fee8ab66db9a.PNG)

   Current Work:
 I can export the Node ID's, X, Y, Z coordinates and the stress points. I have attempted to create the new body based on the X, Y, Z points using pyvista, but I have been experiencing challenges in doing this. 

Simplified Work Flow: 
The image below illustrates a simple plate ( 50mm x 10mm x 2mm ) which I would like to recreate using pyvista for nodes that experience a specific stress value. In this simplified workflow, I am using all the node coordinates to understand how to achieve this using pyvista. 

![image](https://user-images.githubusercontent.com/85831923/121808416-b02f2480-cc58-11eb-8b62-dbc415122e74.png)

I can export the Node ID's, X, Y, Z and Stress Values in a CSV file format. Please see the image and the CSV file attached below. 

![Exporting_CSV](https://user-images.githubusercontent.com/85831923/121808517-30558a00-cc59-11eb-99fe-c5119fd6dd1c.PNG)

The CSV file: [great4.zip](https://github.com/pyvista/pyvista-support/files/6643912/great4.zip)
 
Using PyVista: 

To first understand if my idea is possible, I have been trying to recreate the exact plate using all the coordinate points. The code I have used to do this is as follows: 

```javascript
import pandas as pd
import numpy as np
import pyvista as pv

df = pd.read_csv('great4.csv', delimiter = ';')
coords = df[[' X Coord', ' Y Coord', ' Z Coord']].values

mesh = pv.PolyData(coords)
#mesh.plot(show_grid=True,point_size=5)
surf = mesh.delaunay_3d(alpha=6, tol=0.05)
surf.plot(background='w', show_edges=True, show_grid=False)

pv.save_meshio('mesh.stl', surf)"
```

The code above is able to produce a mesh plot using the delaunay_3d() function see image below: 

![image](https://user-images.githubusercontent.com/85831923/121808867-c5a54e00-cc5a-11eb-8d10-64758998217f.png)

When I try to create an STL file from the mesh, I get an error. 

I also tried to use the structured grid approach to recreate the model, but the result does not seem correct. The code I used for this is illustrated below: 

```javascript
import pandas as pd
import numpy as np
import pyvista as pv
import matplotlib.pyplot as plt

df = pd.read_csv('great4.csv', delimiter = ';')
coords = df[[' X Coord', ' Y Coord', ' Z Coord']].values

structured = pv.StructuredGrid(coords)
structured.points = coords
structured.dimensions = [26, 10, 2]
elevation = structured.elevation()
elevation.plot(show_edges=True, show_grid=True, notebook = True)
pv.save_meshio('mesh.stl', structured)"
```

The image below represents the result of the above code: 
![image](https://user-images.githubusercontent.com/85831923/121809287-8b3cb080-cc5c-11eb-8d06-aeecea334d85.png)

I do apologise if this is a straightforward exercise and if my coding is completely incorrect. I am still learning Python and would like to understand using pyvista for this type of application. 

Some questions: 
1. Does the delaunay_3d() function create a 3d solid triangulation, or is it just creating a surface around my points? 
2. Is it possible to do what I am trying to using pyvista? 

Thank you for taking the time to read through my post, and any help and assistance would be greatly appreciated. 

Kind Regards

**Comment by akaszynski** [comment]
Thanks for posting here.

You're going to want to use the UnstructuredGrid class to represent this FEM.  Here's an example:
https://docs.pyvista.org/examples/00-load/create-unstructured-surface.html

I've been able to apply this with partial success to this data.  It seems that the cell point ordering is inconsistent in this output.  Rather than doing this by hand, see if your FEA program can output it in a consistent format or see if any of the output formats are supported by `meshio`.  If you can read it in with `meshio`, you can visualize it with `pyvista` as we support `meshio` files.

```py
import vtk
import pandas as pd
import numpy as np
import pyvista as pv
import matplotlib.pyplot as plt

df = pd.read_csv('great4.csv', delimiter = ';')


# need to create a unstructuredgrid from this output

coord = np.vstack((df[' X Coord'], df[' Y Coord'], df[' Z Coord'])).T
n_node = coord.shape[0]
elem_ids = np.arange(0, n_node, dtype=np.int32).reshape(-1, 8)
n_elem = elem_ids.shape[0]

# these cells don't follow the VTK hexahedral ordering.  Swap (2, 3), (6, 7)
ids_col2, ids_col3 = elem_ids[:, 2:4].T.copy()
elem_ids[:, 2] = ids_col3
elem_ids[:, 3] = ids_col2

ids_col6, ids_col7 = elem_ids[:, 6:].T.copy()
elem_ids[:, 6] = ids_col7
elem_ids[:, 7] = ids_col6

# do the same with the stress
stress = df['Von-Mises'].to_numpy().reshape(-1, 8)
stress_col2, stress_col3 = stress[:, 2:4].T.copy()
stress[:, 2] = stress_col3
stress[:, 3] = stress_col2

stress_col6, stress_col7 = stress[:, 6:].T.copy()
stress[:, 6] = stress_col7
stress[:, 7] = stress_col6


# add the cell padding
padding = np.ones((n_elem, 1), dtype=np.int64)*8
cells = np.hstack((padding, elem_ids)).ravel()

# cells are all vtk.vtkQuadraticTetra
cell_type = np.empty(n_elem, np.uint8)
cell_type[:] = vtk.VTK_HEXAHEDRON

grid = pv.UnstructuredGrid(cells, cell_type, coord)
grid['stress'] = stress.ravel()

grid.plot(cmap='jet')
```
![sc](https://user-images.githubusercontent.com/11981631/121813660-439d3100-cc2a-11eb-8748-f4bc4d957dc4.png)
